html/modules/custom/ghi_plans/src/Plugin/ParagraphHandler/PlanEntityTypes.php

```diff
@@ -0,0 +1,469 @@
+        '#empty' => $this->t('No suitable plan entities found. If you save this form like this, the block will not be displayed.'),
+      ];
+
+      if (count($matching_entities)) {
+        $validation_options = $defaults;
+        $validation_options['entity_ids'] = NULL;
+        foreach ($matching_entities as $entity) {
+          if ($this->validatePlanEntity($entity, $validation_options)) {
+            continue;
+          }
+          $subform['entity_ids'][$entity->id]['#disabled'] = TRUE;
+        }
+      }
+    }
+  }
+
+  /**
+   * Ajax callback to update the entities list.
+   *
+   * @param array $form
+   *   Form array.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   Form state interface.
+   *
+   * @return mixed
+   *   The part of the form array that should be updated.
+   */
+  public function updateEntitiesList(array &$form, FormStateInterface $form_state) {
+    $triggering_element = $form_state->getTriggeringElement();
+    $parents = $triggering_element['#array_parents'];
+    array_pop($parents);
+    $parents[] = 'entity_ids';
+    return NestedArray::getValue($form, $parents);
+  }
+
+  /**
+   * Get options for the entity type dropdown.
+   *
+   * @return array
+   *   An array with valid options for the current context.
+   */
+  private function getEntityTypeOptions() {
+    $plan_id = $this->parentEntity->field_original_id->value;
+
+    if (!$plan_id) {
+      // Without a plan context, we show a default set of availabe plan entity
+      // types.
+      $default_options = [
+        'CA' => $this->t('Cluster Activities'),
+        'CO' => $this->t('Cluster Objectives'),
+        'CQ' => $this->t('Humanitarian Consequences'),
+        'SO' => $this->t('Strategic Objectives'),
+        'SP' => $this->t('Specific Objectives'),
+        'SSO' => $this->t('Sub Strategic Objectives'),
+        'OC' => $this->t('Outcomes'),
+        'OP' => $this->t('Outputs'),
+      ];
+      return $default_options;
+    }
+
+    $matching_entities = $this->getPlanEntities();
+
+    $options = [];
+    if (empty($matching_entities)) {
+      return $options;
+    }
+    $weight = [];
+    foreach ($matching_entities as $entity) {
+      $ref_code = $entity->entityPrototype->refCode;
+      if (empty($options[$ref_code])) {
+        $name = $entity->entityPrototype->value->name->en->plural;
+        $options[$ref_code] = $name;
+        $weight[$ref_code] = $entity->entityPrototype->orderNumber;
+      }
+    }
+    uksort($options, function ($ref_code_a, $ref_code_b) use ($weight) {
+      return $weight[$ref_code_a] > $weight[$ref_code_b];
+    });
+    return $options;
+  }
+
+  /**
+   * Get available plan entities for the current context.
+   *
+   * @param string $entity_type
+   *   The entity type to restrict the context.
+   *
+   * @return array
+   *   An array of plan entity objects for the current context.
+   */
+  private function getPlanEntities($entity_type = NULL) {
```

Comment:
Move this into a `EndpointQuery` class so it's not tightly coupled to the paragraph and can easily be reused.

Reply:
Hm, I still want to keep the `EndpointQuery` class independent from Insights/GHI specific logic, to be able to properly share it with FTS Public. I already shouldn't have added `ApiEntityHelper` to the `hpc_api` module. From my point of view, those should better live in a GHI specific module.

Reply:
Got you, `EndpointQuery` is generic, but we can add a `GhiEndpointQuery` where we can add GHI specific code, since at some point in the future the API will change. So if `GhiEndpointQuery` returns structured data we only need to refactor the queries in that class.

Best to refactor this asap so impact on other parts is minimal?

Reply:
Fully agreed!

Reply:
I added `PlanEntitiesQuery` for the moment to the `ghi_plans` module. We can add several of those query classes depending on the need.
I would also remove `EndpointPlanQuery` from the `hpc_api` module, as I don't think it will be used, and if so it should also be moved somewhere else and probably renamed.